### PR TITLE
Gas metering for cosmwasm instructions

### PR DESCRIPTION
**Pull request by aeryz**
## Issue

https://app.clickup.com/t/3px83ne

## Checklist

- [X] Updated the rust/typescript docs.

### CODE REVIEW

File: code/parachain/frame/cosmwasm/src/instrument.rs [hunk 1]
```diff
@@ -31,3 +36,299 @@ pub fn gas_and_stack_instrumentation(
...
+			Instruction::GrowMemory(_) => self.growmemory,
+			Instruction::Br(_) => self.br,
+			Instruction::BrIf(_) => self.brif,
+			Instruction::BrTable(table) => self
+				.brtable
+				.saturating_add(self.brtable_per_elem.saturating_mul(table.table.len() as u32)),
+			Instruction::Call(_) => self.call,
+			// TODO(aeryz): CallIndirect will be implemented after the pallet is upgraded to IBC
+			// compatible VM
+			Instruction::CallIndirect(_, _) => self.call,
```
Comment (hussein-aitlahcen) [inline code/parachain/frame/cosmwasm/src/instrument.rs:323]: I don't understand the requirement for IBC here, could you explain :)

Comment (aeryz) [reply code/parachain/frame/cosmwasm/src/instrument.rs:323]: I need to change code_gen but the IBC changes are merged to main. We can either wait for IBC pr in pallet side, or in this PR, I can implement that functions by returning Error::Unsupported

File: code/parachain/frame/cosmwasm/src/runtimes/wasmi.rs [hunk 2]
```diff
@@ -102,12 +102,14 @@ pub enum InitialStorageMutability {
...
 
 /// VM shared cache
+#[derive(Clone)]
 pub struct CosmwasmVMCache {
 	/// Code cache, a mapping from a CosmWasm contract code id to the baking code.
 	pub code: BTreeMap<CosmwasmCodeId, Vec<u8>>,
 }
 
 /// VM shared state
+#[derive(Clone)]
```
Comment (dzmitry-lahoda) [inline code/parachain/frame/cosmwasm/src/runtimes/wasmi.rs:112]: will it clone whole wasm module if it is loaded into?  is it too risky to miss such code in production?

File: code/parachain/frame/cosmwasm/src/weights.rs [hunk 3]
```diff
@@ -61,6 +62,71 @@ pub trait WeightInfo {
 	fn query_info() -> Weight;
```
[the remark this answers is not in the document]
Comment (aeryz) [reply code/parachain/frame/cosmwasm/src/weights.rs:62]: Actually, I used a template file which I got from polkadot to generate weights in this format. I think we can agree on a file like this and put it to the repository.

[the remark this answers is not in the document]
Comment (dzmitry-lahoda) [reply code/parachain/frame/cosmwasm/src/weights.rs:62]: cool. how you generated weights?

[the remark this answers is not in the document]
Comment (dzmitry-lahoda) [reply code/parachain/frame/cosmwasm/src/weights.rs:62]: per instruction weight.

**Pull request merged**